### DISCUSSION TITLE
Calculation of pk2_jacobian in moose

### DISCUSSION THREAD

**Discussion by manikumd22** [idaholab/moose#29111] [Q&A Modules: Solid mechanics]
Hi @GiudGiud ,

In MOOSE, when calculating the PK2 Jacobian (C=∂S/∂E), does the framework use automatic differentiation? Specifically, if I directly compute the Second Piola-Kirchhoff stress (S) in my model, will MOOSE automatically differentiate S with respect to the Green-Lagrange strain (E) to compute the Jacobian (C)?

**Comment by maxnezdyur** [comment]
No that system does not use AD. It does have perfect handcoded jacobians. I am not sure what you mean by "directly computing" S. If you changed how you compute S then I think you will need to fix the pk2 jacobian too.

**Reply by manikumd22** [reply]
Hi @maxnezdyur 

For instance, for a neo-Hookean material, we compute it in this way:


```
 // Large deformation = nonlinear strain
if (_large_kinematics)
{
  RankTwoTensor Cinv = (2 * _E[_qp] + RankTwoTensor::Identity()).inverse();
  _S[_qp] = (_lambda[_qp] * log(_F[_qp].det()) - _mu[_qp]) * Cinv +
            _mu[_qp] * RankTwoTensor::Identity();
  _C[_qp] =
      -2 * (_lambda[_qp] * log(_F[_qp].det()) - _mu[_qp]) * Cinv.times<i_, k_, l_, j_>(Cinv) +
      _lambda[_qp] * Cinv.times<i_, j_, k_, l_>(Cinv);
}

```
In my case, I am providing _S_{qp} using a material model that computes _S_{qp} directly, but I did not provide any means to calculate _C_{qp}.

For example, I am doing it like this:
```
// Large deformation = nonlinear strain
if (_large_kinematics)
{
  RankTwoTensor Cinv = (2 * _E[_qp] + RankTwoTensor::Identity()).inverse();
  _S[_qp] = function(_F[_qp]);
}
```

The model is running, but I want to understand how _C_{qp} is being calculated in this case. Are they using some method, such as taking the derivative of _S_{qp} with respect to strain?

In other words, is _C_{qp} being calculated automatically by MOOSE as a function of _S_{qp}?

**Reply by maxnezdyur** [reply]
No I don't believe so. I would look at [AnalyzeJacobian](https://mooseframework.inl.gov/python/testers/AnalyzeJacobian.html), which is give you a way to check if the Jacobian is still accurate.

**Reply by manikumd22** [reply]
Hi @maxnezdyur 

Thanks for your response. I just want to clarify:  is MOOSE automatically using finite-difference approximations to compute the Jacobian (_C_{qp}) in this case?

Thanks in advance!

In the mean time, I am checking the link you shared

**Reply by maxnezdyur** [reply]
I don't see any finite differencing of C in any of the code.

**Reply by manikumd22** [reply]
Hi @maxnezdyur 

I am still confused then which mechanism is moose using to compute the pk2_jacobian


As for calculating the pk1_jacobian they have two expressions one involve S(2nd piola) and other involve C that is called as pk2_jacobian

This is ComputeLagrangianStressPK1
```
  if (_large_kinematics)
  {
    _pk1_stress[_qp] = _F[_qp] * _S[_qp];
    usingTensorIndices(i_, j_, k_, l_);
    RankFourTensor dE =
        0.5 * (RankTwoTensor::Identity().times<i_, l_, j_, k_>(_F[_qp].transpose()) +
               _F[_qp].transpose().times<i_, k_, j_, l_>(RankTwoTensor::Identity()));

    _pk1_jacobian[_qp] = RankTwoTensor::Identity().times<i_, k_, j_, l_>(_S[_qp].transpose()) +
                         (_C[_qp] * dE).singleProductI(_F[_qp]);
  }
```
As I’m just inputting the  _S[_qp] using my material model and the model is running (converging) that means it must be calculating _C[_qp] somehow


As you mentioned moose is not using AD, finite difference

So, which technique is using to compute _C[_qp]

**Reply by GiudGiud** [reply]
You are talking about the LagrangianStress materials and UOs right?
These have coded the analytical expressions for the Jacobians. No AD or finite diffewrence.

**Reply by manikumd22** [reply]
Hi @GiudGiud 

I am talking about LagrangianStress

I looked for the expression where they are calculating the pk2_jacobian but was unable to find it, although i found for pk1_jacobian which is using _C

ComputeLagrangianStressPK1.C
ComputeLagrangianStressPK2.C


Can you please let me know the file name of the expression where moose is explicitly calculating the pk2_jacobian

**Reply by GiudGiud** [reply]
It should be this one
materials/lagrangian/ComputeLagrangianStressPK2.C
as the property is declared there
```
 _C(declareProperty<RankFourTensor>(_base_name + "pk2_jacobian"))
```
then computed

**Reply by manikumd22** [reply]
Hi @GiudGiud 

If you see this expression 

  if (_large_kinematics)
  {
    _pk1_stress[_qp] = _F[_qp] * _S[_qp];
    usingTensorIndices(i_, j_, k_, l_);
    RankFourTensor dE =
        0.5 * (RankTwoTensor::Identity().times<i_, l_, j_, k_>(_F[_qp].transpose()) +
               _F[_qp].transpose().times<i_, k_, j_, l_>(RankTwoTensor::Identity()));

    _pk1_jacobian[_qp] = RankTwoTensor::Identity().times<i_, k_, j_, l_>(_S[_qp].transpose()) +
                         (_C[_qp] * dE).singleProductI(_F[_qp]);
  }
  
I’m only inputting _S[_qp]  not _C[_qp] 

And my model running fine

Does that means it is assuming it to be zero tensor?

**Reply by GiudGiud** [reply]
> I’m only inputting _S[_qp] not _C[_qp]

> And my model running fine

> Does that means it is assuming it to be zero tensor?

no it means something is computing _C and it's somewhere in the derived class not in ComputeLagrangianStressPK2 where
`computeQpPK2Stress()` is a pure virtual. 
For example in ComputeNeoHookeanStress you can find this computation

**Reply by manikumd22** [reply]
Hi @GiudGiud 

Got it

Correct me if I’m wrong


It means indirectly it will be using _S somehow to compute?


As generally _C calculates as derivative of _S for instance after taking derivative of _S in neohookean that analytic expression comes

**Reply by GiudGiud** [reply]
_S is also computed in the derived class
```
    _S[_qp] = (_lambda[_qp] * log(_F[_qp].det()) - _mu[_qp]) * Cinv +
              _mu[_qp] * RankTwoTensor::Identity();
    _C[_qp] =
        -2 * (_lambda[_qp] * log(_F[_qp].det()) - _mu[_qp]) * Cinv.times<i_, k_, l_, j_>(Cinv) +
        _lambda[_qp] * Cinv.times<i_, j_, k_, l_>(Cinv);
```
_S is actually not used for _C here, but you can find both _S and _C directly used to compute pk1
```
_pk1_jacobian[_qp] = RankTwoTensor::Identity().times<i_, k_, j_, l_>(_S[_qp].transpose()) +
                     (_C[_qp] * dE).singleProductI(_F[_qp]);
```

**Reply by manikumd22** [reply]
Hi @GiudGiud 

For my case I’m using my material model to obtain the S and not calculating the C

as you mentioned earlier that 

no it means something is computing _C and it's somewhere in the derived class not in ComputeLagrangianStressPK2 where
computeQpPK2Stress() is a pure virtual.


Wherever it is calculating C, will it be using S?

**Reply by GiudGiud** [reply]
> Wherever it is calculating C, will it be using S?

That depends on the implementation of `computeQpPK2Stress`

do you define that routine yourself? How did you code it?

**Reply by manikumd22** [reply]
Hi @GiudGiud 

I defined my material in this way, I'm modifying by coding in this way

ComputeLinearRandom.C


void
ComputeLinearRandom::computeQpPK2Stress()
{
  if (_large_kinematics)
 { 
        for (int pp = 0; pp < 24; ++pp) {     
            if (MooseUtils::absoluteFuzzyEqual(_q_point[_qp](0), _quad_points[pp]))
 {           

...
...
                _weight_input_layer_5_h1[_qp].resize(64, 1);
                for (int i = 0; i < 64; ++i) {
 Real value = _weight_output_layer_4_h1[_qp](i, 0);
                    _weight_input_layer_5_h1[_qp](i, 0) = std::max(value, 0.0); 
 }  

                _weight_output_layer_5_h1[_qp].resize(3, 1);
                for (int i = 0; i < 3; ++i) {
                    _weight_output_layer_5_h1[_qp](i, 0) = 0.0;
                    for (int j = 0; j < 64; ++j) {
                        _weight_output_layer_5_h1[_qp](i, 0) += _weight_layer_5_h1[_qp](i, j) * _weight_input_layer_5_h1[_qp](j, 0);
 }
                    _weight_output_layer_5_h1[_qp](i, 0) = (-1.15513 * std::pow(_stiff[pp], 4) + 2.77110 * std::pow(_stiff[pp], 3)  - 2.03973 * std::pow(_stiff[pp], 2)  + 1.09031 * _stiff[pp]  + 0.74489)*(_weight_output_layer_5_h1[_qp](i, 0) + _weight_bias_5_h1[_qp](i, 0));
 }


                // // Homogeneous
                _weight_output_layer_5_h1[_qp](0,0) = (_weight_output_layer_5_h1[_qp](0,0)*(0.0057805 + 1e-4) + 2.49990627e-02);
                _weight_output_layer_5_h1[_qp](1,0) = (_weight_output_layer_5_h1[_qp](1,0)*(0.0005845 + 1e-4) - 1.92484876e-11);
                _weight_output_layer_5_h1[_qp](2,0) = (_weight_output_layer_5_h1[_qp](2,0)*(0.0057805 + 1e-4) + 2.49990627e-02);


                _S[_qp](0,0) = 1000000*_weight_output_layer_5_h1[_qp](0,0);
                _S[_qp](0,1) = 10*_weight_output_layer_5_h1[_qp](1,0);
                _S[_qp](1,0) = 10*_weight_output_layer_5_h1[_qp](1,0);
                _S[_qp](1,1) = 1000*_weight_output_layer_5_h1[_qp](2,0);


Do I have to modify something else as well?

**Reply by GiudGiud** [reply]
You should be computing _C too since it is used later on

**Reply by manikumd22** [reply]
Hi @GiudGiud 

I'm not defining any explicit analytic expression to compute _C

I'm not able ti find how moose is computing _C as _C is required in the one of term to evaluate the pk1_jacobian

**Reply by manikumd22** [reply]
Hi @GiudGiud 

I can compute the _C[_qp] using _C[_qp] = partial (_S[_qp])/partial (_E[_qp]) after computing this

            _S[_qp](0,0) = 1000000*_weight_output_layer_5_h1[_qp](0,0);
            _S[_qp](0,1) = 10*_weight_output_layer_5_h1[_qp](1,0);
            _S[_qp](1,0) = 10*_weight_output_layer_5_h1[_qp](1,0);
            _S[_qp](1,1) = 1000*_weight_output_layer_5_h1[_qp](2,0);



Could you please tell me how to calculate this derivative using any means to compute?

**Reply by GiudGiud** [reply]
do you know the dependency of weight_output_layer_5_h1 on E ?

**Reply by manikumd22** [reply]
Hi @GiudGiud 

weight_output_layer_5_h1 is function of _F

<img width="341" alt="Screenshot 2024-11-23 at 12 48 53 AM" src="https://github.com/user-attachments/assets/6b89cf1e-d48f-43f6-b212-1f8370d2b6eb">

If somehow I got derivative of S wrt F then I will be able to get the C

**Reply by GiudGiud** [reply]
I see. Could you use finite differencing there to compute it at F and F+ epsilon and get an approximate derivative ?

**Reply by manikumd22** [reply]
Hi @GiudGiud  

I obtained _F[_qp], and perturbed_deformation, and similarly I obtained the _S[qp] and perturbed_stress
```
 // Create a copy of the current deformation gradient tensor
  RankTwoTensor perturbed_deformation = _F[_qp];

  // Add a small perturbation delta
  Real delta = 1e-8; // Small perturbation value

  perturbed_deformation(0, 0) += delta; // Add delta to the (k, l) component
  perturbed_deformation(0, 1) += delta; // Add delta to the (k, l) component
  perturbed_deformation(1, 0) += delta; // Add delta to the (k, l) component
  perturbed_deformation(1, 1) += delta; // Add delta to the (k, l) component


                // Create a copy of the current deformation gradient tensor
_perturbed_stress[_qp].resize(2, 2);


_perturbed_stress[_qp](0,0) = 1000000*_weight_output_layer_5_h1_new[_qp](0,0);
_perturbed_stress[_qp](0,1) = 10*_weight_output_layer_5_h1_new[_qp](1,0);
_perturbed_stress[_qp](1,0) = 10*_weight_output_layer_5_h1_new[_qp](1,0);
_perturbed_stress[_qp](1,1) = 1000*_weight_output_layer_5_h1_new[_qp](1,1);
```

Can you please tell me how to do the finite difference?
```
 RankFourTensor _C[_qp] = ??
```
Is it the correct way to perturb the _F components, as I am perturbing by a tiny amount equal for each component?

**Reply by GiudGiud** [reply]
@reverendbedford will know more than me on how to do this. 
I dont see the link between weight_output_layer_5_h1_new and _F so I don't see the link between _F and _S

**Reply by manikumd22** [reply]
Hi @GiudGiud 

I didn't write the full equations that relate the  _perturbed_deformation with weight_output_layer_5_h1_new

I did the calculation for _C

Is it correct way to do the finite difference?

RankTwoTensor original_F = _F[_qp]; // Original deformation gradient
RankTwoTensor original_S = _S[_qp]; // Original stress tensor
RankFourTensor C; // Fourth-order elasticity tensor to store results

// Loop over all components of F
for (unsigned int k = 0; k < 3; ++k) // Assume 3x3 tensor for 3D
{
    for (unsigned int l = 0; l < 3; ++l)
    {
        // Create a perturbed deformation gradient
        RankTwoTensor perturbed_F = original_F;
        perturbed_F(k, l) += delta;

        // Compute the perturbed stress tensor
        RankTwoTensor perturbed_S;
        perturbed_S(0, 0) = 1000000 * _weight_output_layer_5_h1_new[_qp](0, 0);
        perturbed_S(0, 1) = 10 * _weight_output_layer_5_h1_new[_qp](0, 1);
        perturbed_S(1, 0) = 10 * _weight_output_layer_5_h1_new[_qp](1, 0);
        perturbed_S(1, 1) = 1000 * _weight_output_layer_5_h1_new[_qp](1, 1);

        // Compute finite difference derivative for each component of S
        for (unsigned int i = 0; i < 3; ++i) // Rows of S
        {
            for (unsigned int j = 0; j < 3; ++j) // Columns of S
            {
                // Compute C_{ijkl} = (S_ij(perturbed) - S_ij(original)) / delta
                C(i, j, k, l) = (perturbed_S(i, j) - original_S(i, j)) / delta;
            }
        }
    }
}

// Store the computed elasticity tensor (C) for further use
_C[_qp] = C;

**Reply by reverendbedford** [reply]
As @GiudGiud says this set of kernel's doesn't do AD, you'll need to provide the derivative $\frac{dS_{ij}}{dE_{kl}}$.

Given that you seem to know $\frac{dS_{ij}}{dF_{kl}}$ it may be easier to just convert your PK2 stress to PK1 and use the [PK1](https://mooseframework.inl.gov/source/materials/lagrangian/ComputeLagrangianStressPK1.html) object.  That would take care of the chain rule for you.

Are you sure your model is well formed?  The PK2 stress has to be symmetric, so it's odd to see derivatives with respect to the unsymmetric deformation gradient turning up.

**Reply by manikumd22** [reply]
Hi @reverendbedford 

For C I need to provide the partial P/ partial F, as I am already calculating the S using my material model, I, and calculate the P using P =  FS

and then take the derivative wrt F

For that case, do you know if PK2 object will work?
```
  // Complicated wrapping, see documentation
  if (_large_kinematics)
  {
    _pk1_stress[_qp] = _F[_qp] * _S[_qp];
    usingTensorIndices(i_, j_, k_, l_);
    RankFourTensor dE =
        0.5 * (RankTwoTensor::Identity().times<i_, l_, j_, k_>(_F[_qp].transpose()) +
               _F[_qp].transpose().times<i_, k_, j_, l_>(RankTwoTensor::Identity()));

    _pk1_jacobian[_qp] = RankTwoTensor::Identity().times<i_, k_, j_, l_>(_S[_qp].transpose()) +
                         (_C[_qp] * dE).singleProductI(_F[_qp]);
  }
  // Small deformations all are equivalent
  else
  {
    _pk1_stress[_qp] = _S[_qp];
    _pk1_jacobian[_qp] = _C[_qp];
  }

```

I modified my method to calculate C as follows:
```
                RankTwoTensor original_F = _F[_qp]; // Original deformation gradient
                RankTwoTensor original_S = _S[_qp]; // Original stress tensor
                RankFourTensor C; // Fourth-order elasticity tensor to store results

                // Loop over all components of F
                for (unsigned int k = 0; k < 2; ++k) // Assume 2x2 for 2D (adjust for 3D as needed)
                {
                    for (unsigned int l = 0; l < 2; ++l)
                    {
                        // Perturb the (k, l) component of F
                        RankTwoTensor perturbed_F = original_F;
                        perturbed_F(k, l) += delta;

                        // Compute original P
                        RankTwoTensor original_P;
                        original_P(0, 0) = original_F(0, 0) * _S[_qp](0, 0) + original_F(0, 1) * _S[_qp](0, 1);
                        original_P(0, 1) = original_F(0, 0) * _S[_qp](0, 1) + original_F(0, 1) * _S[_qp](1, 1);
                        original_P(1, 0) = original_F(1, 0) * _S[_qp](0, 0) + original_F(1, 1) * _S[_qp](1, 0);
                        original_P(1, 1) = original_F(1, 0) * _S[_qp](0, 1) + original_F(1, 1) * _S[_qp](1, 1);

                        // Compute perturbed P
                        RankTwoTensor perturbed_P;
                        perturbed_P(0, 0) = perturbed_F(0, 0) * _perturbed_stress[_qp](0, 0) + perturbed_F(0, 1) * _perturbed_stress[_qp](0, 1);
                        perturbed_P(0, 1) = perturbed_F(0, 0) * _perturbed_stress[_qp](0, 1) + perturbed_F(0, 1) * _perturbed_stress[_qp](1, 1);
                        perturbed_P(1, 0) = perturbed_F(1, 0) * _perturbed_stress[_qp](0, 0) + perturbed_F(1, 1) * _perturbed_stress[_qp](1, 0);
                        perturbed_P(1, 1) = perturbed_F(1, 0) * _perturbed_stress[_qp](0, 1) + perturbed_F(1, 1) * _perturbed_stress[_qp](1, 1);

                        // Compute finite difference derivative
                        for (unsigned int i = 0; i < 2; ++i) // Rows of P
                        {
                            for (unsigned int j = 0; j < 2; ++j) // Columns of P
                            {
                                // C_{ijkl} = (P_{ij}(perturbed) - P_{ij}(original)) / delta
                                C(i, j, k, l) = (perturbed_P(i, j) - original_P(i, j)) / delta;
                            }
                        }
                    }
                }
```

**Reply by manikumd22** [reply]
Hi @GiudGiud 

I have a doubt

If I use pk1 kernel 

Then only P is required

If I use pl2 kernel then I have to provide both S and C tensor

Is that correct?

**Reply by GiudGiud** [reply]
You mean the pk1 and pk2 lagrangian stress materials? Or which kernels specifically?

**Reply by manikumd22** [reply]
Hi @GiudGiud 

Kernels

For instance of you solve the neohookean stress it uses pk2 kernel 

Define both the S and C

**Reply by GiudGiud** [reply]
then yes. The PK2 materials are defining both S and C
```
ComputeLagrangianStressPK2::ComputeLagrangianStressPK2(const InputParameters & parameters)
  : ComputeLagrangianStressPK1(parameters),
    _E(declareProperty<RankTwoTensor>(_base_name + "green_lagrange_strain")),
    _S(declareProperty<RankTwoTensor>(_base_name + "pk2_stress")),
    _C(declareProperty<RankFourTensor>(_base_name + "pk2_jacobian"))
```
so you need to do that too

**Reply by manikumd22** [reply]
Hi @GiudGiud 

For pk1 kernel

Only P will be used to calculate the jacobian

Right?

**Reply by reverendbedford** [reply]
The way all these materials work is they provide you some strain measure and expect you to return:

1. A stress measure
2. The derivative of that stress measure with respect to the strain measure.

The material system does all the other work for you to get the correct consistent tangent/Jacobian for the kernel proper.

For the PK1 material you get the deformation gradient and you are expected to return the PK1 stress and the derivative of the PK1 stress with respect to the deformation gradient.

For the PK2 you get the Green-Lagrange strain and you are expected to return the PK2 stress and the derivative of the PK2 stress with respect to the Green-Lagrange strain.

You can pick either and get consistent/correct results.  It's just an aesthetic choice: what's easiest for you to do for your particular constitutive model.

**Reply by manikumd22** [reply]
Hi @GiudGiud 

After obtaining the stress and taking derivative of stress with respect to strain or F depending on the kernel

For pk1 kernel

Derivative of stress with respect F is also _C[_qp]

As for the pk2 kernel is is like that

**Reply by reverendbedford** [reply]
The material objects will calculate the strain, all you have to do is calculate the stress and derivative.

For PK1 we don't have a "native" material model using this interface directly.  But the PK2 actually derives from it, so you can see what you need to set [here](https://github.com/idaholab/moose/blob/next/modules/solid_mechanics/src/materials/lagrangian/ComputeLagrangianStressPK2.C).  You "get" the deformation gradient and need to provide the PK1 stress and the derivative of PK1 with respect to the deformation gradient.

For PK2 there is a complete example [here](https://github.com/idaholab/moose/blob/next/modules/solid_mechanics/src/materials/lagrangian/ComputeStVenantKirchhoffStress.C).  You "get" the Green-Lagrange strain and have to provide the PK2 stress and the derivative of PK2 with respect to the deformation gradient.

**Reply by manikumd22** [reply]
Hi @GiudGiud 


As for the Pk1 kernel
this is formulation

If I calculate _pk1_stress[_qp], then where it is taking derivative?

  if (_large_kinematics)
  {
    _cauchy_stress[_qp] = _pk1_stress[_qp] * _F[_qp].transpose() / _F[_qp].det();

    auto f = _inv_df[_qp].inverse();
    usingTensorIndices(i_, j_, k_, l_);
    _cauchy_jacobian[_qp] = _cauchy_stress[_qp].times<i_, l_, j_, k_>(f) -
                            _cauchy_stress[_qp].times<i_, j_, k_, l_>(f.transpose());
    _cauchy_jacobian[_qp] +=
        _pk1_jacobian[_qp].tripleProductJkl(_F[_qp], f.transpose(), _F[_qp]) / _F[_qp].det();
  }

**Reply by reverendbedford** [reply]
You don't worry about the kernel, you just need to worry about the material.  The rest of the system takes care of calculating the derivative the kernel needs.

So if you want to inherit from the `ComputeLagrangianStressPK1` object all your material needs to do is set:
- `_pk1_stress[_qp]` to the PK1 stress
- `_pk1_jacobian[_qp]` to the derivative of the PK1 stress  with respect to the deformation gradient.

This will happen in your material object, which again should derive from `ComputeLagrangianStressPK1`

**Reply by manikumd22** [reply]
Hi @reverendbedford 

I have calculated the _pk1stress and perturbed _pk1 stress 

this is how I will calculate the _pk1jacobian[_qp], right? 


                RankTwoTensor original_F = _F[_qp]; // Original deformation gradient
                RankTwoTensor original_P = _pk1_stress[_qp]; // Original stress tensor
                RankTwoTensor perturbed_P = _perturbed_P[_qp]; // Original stress tensor
                RankFourTensor C; // Fourth-order elasticity tensor to store results

                // Loop over all components of F
                for (unsigned int k = 0; k < 2; ++k) // Assume 2x2 for 2D (adjust for 3D as needed)
                {
                    for (unsigned int l = 0; l < 2; ++l)
                    {
                        // Perturb the (k, l) component of F
                        RankTwoTensor perturbed_F = original_F;
                        perturbed_F(k, l) += delta;

                        // Compute finite difference derivative
                        for (unsigned int i = 0; i < 2; ++i) // Rows of P
                        {
                            for (unsigned int j = 0; j < 2; ++j) // Columns of P
                            {
                                // C_{ijkl} = (P_{ij}(perturbed) - P_{ij}(original)) / delta
                                C(i, j, k, l) = (perturbed_P(i, j) - original_P(i, j)) / (delta*10e9);
                            }
                        }
                    }
                }

                _pk1_jacobian[_qp] = C;

**Reply by reverendbedford** [reply]
Uhh, this is something you'll probably have to debug in detail yourself, but a couple of things stand out:

1. All MOOSE (solid mechanics) materials should be setup in 3D, even if your simulation is 2D or 1D.
2. The idea of perturbing the deformation gradient is right, but where are you actually calling your model again to get the perturbed PK1 stress?
3. Not sure where the magic `10e9` is coming from here.
